presto-main/src/main/java/com/facebook/presto/operator/project/MergingPageOutput.java

```diff
@@ -0,0 +1,175 @@
+        this(types, DEFAULT_MIN_PAGE_SIZE_IN_BYTES, DEFAULT_MIN_ROW_COUNT, DEFAULT_MAX_PAGE_SIZE_IN_BYTES);
+    }
+
+    public MergingPageOutput(Iterable<? extends Type> types, int minPageSizeInBytes, int minRowCount, int maxPageSizeInBytes)
+    {
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.minPageSizeInBytes = minPageSizeInBytes;
+        this.minRowCount = minRowCount;
+        pageBuilder = PageBuilder.withMaxPageSize(maxPageSizeInBytes, this.types);
+    }
+
+    public boolean needsInput()
+    {
+        return !finishing && outputQueue.isEmpty() && inputQueue.isEmpty();
+    }
+
+    public void addInput(PageProcessorOutput input)
+    {
+        requireNonNull(input, "input is null");
+        checkState(!finishing, "output is in finishing state");
+        inputQueue.add(input);
+    }
+
+    @Nullable
+    public Page getOutput()
+    {
+        if (!outputQueue.isEmpty()) {
+            return outputQueue.poll();
+        }
+
+        while (!inputQueue.isEmpty() && outputQueue.isEmpty()) {
+            Iterator<Page> input = inputQueue.peek();
+            if (input.hasNext()) {
+                process(input.next());
+            }
+            if (!input.hasNext()) {
+                inputQueue.poll();
+            }
+        }
+
+        if (inputQueue.isEmpty() && finishing) {
+            flush();
+        }
+
+        return outputQueue.poll();
+    }
+
+    public void finish()
+    {
+        finishing = true;
+    }
+
+    public boolean isFinished()
+    {
+        return finishing && inputQueue.isEmpty() && outputQueue.isEmpty() && pageBuilder.isEmpty();
+    }
+
+    private void process(Page page)
+    {
+        requireNonNull(page, "page is null");
+
+        // avoid memory copying for pages that are big enough
+        if (page.getSizeInBytes() >= minPageSizeInBytes || page.getPositionCount() >= minRowCount) {
+            flush();
+            outputQueue.add(page);
+            return;
+        }
+
+        buffer(page);
+    }
+
+    private void buffer(Page page)
+    {
+        for (int position = 0; position < page.getPositionCount(); position++) {
+            pageBuilder.declarePosition();
+            for (int channel = 0; channel < types.size(); channel++) {
+                Type type = types.get(channel);
+                type.appendTo(page.getBlock(channel), position, pageBuilder.getBlockBuilder(channel));
+            }
+        }
+        if (pageBuilder.isFull()) {
+            flush();
+        }
+    }
+
+    private void flush()
+    {
+        if (!pageBuilder.isEmpty()) {
+            Page output = pageBuilder.build();
+            pageBuilder.reset();
+            outputQueue.add(output);
+        }
+    }
+
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE +
+                pageBuilder.getRetainedSizeInBytes() +
+                inputQueue.stream().mapToLong(PageProcessorOutput::getRetainedSizeInBytes).sum() +
```

Comment:
It is not trivial to run count on a go. `PageProcessorOutput` has `getRetainedSize` method implemented in not trivial way. In fact there shouldn't be more than 2 items in those queues. So, i don't think it will be considerably more expensive than it was before.